File: src/main/java/com/example/dpop/orchestrator/account/AccountBindingKeyMappingService.java
package com.example.dpop.orchestrator.account;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.Optional;

@Service
public class AccountBindingKeyMappingService {

    private final AccountBindingKeyMappingRepository repository;

    public AccountBindingKeyMappingService(AccountBindingKeyMappingRepository repository) {
        this.repository = repository;
    }

    @Transactional
    public void mapBindingKeyToAccount(String bindingKeyRef, Long accountId) {
        repository.findByBindingKeyRef(bindingKeyRef).ifPresent(repository::delete);
        repository.save(new AccountBindingKeyMapping(bindingKeyRef, accountId, Instant.now()));
    }

    @Transactional(readOnly = true)
    public Optional<Long> findAccountIdByBindingKeyRef(String bindingKeyRef) {
        return repository.findByBindingKeyRef(bindingKeyRef).map(AccountBindingKeyMapping::getAccountId);
    }
}

File: src/main/kotlin/com/example/dpop/orchestrator/account/AccountBindingKeyMappingService.kt
package com.example.dpop.orchestrator.account

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant

@Service
class AccountBindingKeyMappingService(private val repository: AccountBindingKeyMappingRepository) {

    @Transactional
    fun mapBindingKeyToAccount(bindingKeyRef: String, accountId: Long) {
        repository.findByBindingKeyRef(bindingKeyRef)?.let(repository::delete)
        repository.save(AccountBindingKeyMapping(bindingKeyRef, accountId, Instant.now()))
    }

    @Transactional(readOnly = true)
    fun findAccountIdByBindingKeyRef(bindingKeyRef: String): Long? =
        repository.findByBindingKeyRef(bindingKeyRef)?.accountId
}
